File: Assets/Scripts/Manager/UIManager.cs
using System.Collections.Generic;
using UnityEngine;

///
/// UI管理器，统一管理游戏中会出现的各类UI
///
public class UIManager
{
    //单例
    private static UIManager instance;

    private Transform uiRoot;

    //路径配置字典
    public Dictionary<string, string> pathDict;
    //预制件缓存字典
    public Dictionary<string, GameObject> prefabDict;
    //已打开界面缓存
    public Dictionary<string, BasePanel> panelDict;
    //已打开界面堆栈
    public Stack<BasePanel> panelStack;

    private UIManager()
    {
        prefabDict = new Dictionary<string, GameObject>();
        panelDict = new Dictionary<string, BasePanel>();
        pathDict = new Dictionary<string, string>()
        {
            { UIConst.HeroBackPack,"/HeroBackPack"},
            { UIConst.DrawCard,"/DrawCard"},
            { UIConst.MainMenu,"/MainMenu"}
        };
        panelStack = new Stack<BasePanel>();

    }
    public static UIManager Instance
    {
        get
        {
            if (instance == null)
            {
                instance = new UIManager();
            }
            return instance;
        }
    }

    public Transform UIRoot
    {
        get
        {
            if (uiRoot == null)
            {
                if (GameObject.Find("Canvas"))
                {
                    uiRoot = GameObject.Find("Canvas").transform;
                }
                else
                {
                    uiRoot = new GameObject("Canvas").transform;
                }

            }
            return uiRoot;
        }
    }

    //打开界面
    public void OpenPanel(string name)
    {
        //已经打开的界面不打开
        BasePanel panel = null;
        if (panelDict.TryGetValue(name, out panel))
        {
            return;
        }

        //找不到界面也不打开
        string path = "";
        if (!pathDict.TryGetValue(name, out path))
        {
            return;
        }

        //设置下层界面不可见
        if (panelStack.Count != 0)
        {
            panelStack.Peek().SetOnShow(false);
        }

        //生成对应控制器
        UIFactory factory = new UIFactory();
        factory.CreateController(name);
    }

    //关闭界面
    public bool ClosePanel(string name)
    {
        //没打开的不关
        BasePanel panel = null;
        if (!panelDict.TryGetValue(name, out panel))
        {
            return false;
        }

        //关闭对应界面
        BasePanel nowPanel= panelStack.Pop();
        nowPanel.ClosePanel();

        //设置下层界面可见
        if (panelStack.Count != 0)
        {
            panelStack.Peek().SetOnShow(true);
        }

        //移除缓存
        if (panelDict.ContainsKey(name))
        {
            panelDict.Remove(name);
        }

        return true;
    }
}

public class UIConst
{
    public const string HeroBackPack = "HeroBackPack";
    public const string DrawCard = "DrawCard";
    public const string MainMenu = "MainMenu";
} 

public class UIFactory
{
    public void CreateController(string name)
    {
        switch (name)
        {
            case UIConst.MainMenu:
                MainPageController mainPageController = new MainPageController();
                break;

            case UIConst.HeroBackPack:
                BagController bagController = new BagController();
                break;

            case UIConst.DrawCard:
                CardController cardController = new CardController();
                break;
            default:
                break;
        }
    }
}

File: Assets/LuaScripts/UI/UIManager.lua
UIManager = BaseClass("UIManager")
-- 单例类
function UIManager:Instance()  
    if self.instance == nil then  
        self.instance = self:New()  
    end  
    return self.instance  
end  

-- 初始化函数
function UIManager:__init(params)
    ---初始化参数
    --根ui
    self.uiRoot=GameObject.Find("Canvas").transform
    --页面控制器缓存字典
    self.controllerDict={}
    --已打开页面列表 只存名字
    self.panelList={}
    --打开中的界面栈
    self.panelStack=Stack:Create()
end

-- 打开界面
--打开界面方法
function UIManager:OpenUI(name)
    -- 已打开的界面不打开
    if TableUtil.ContainKeys(self.panelList,name) then
        return
    end
    -- 先找缓存 没有再加载
    if TableUtil.ContainKeys(self.controllerDict,name) then
        self.controllerDict[name]:ShowView()
    else
        self.controllerDict[name]=UIConfigDic[name].controller:New(name)
    end
    --刷新界面
    self.controllerDict[name]:RefreshUI()
    --下层界面不可见
    if self.panelStack:Peek()~=nil then
        self.panelStack:Peek():CloseView()
    end
    --压栈 加入已打开列表
    self.panelStack:Push(self.controllerDict[name])
    table.insert(self.panelList,name)
end

--关闭界面
function UIManager:CloseUI()
    -- 没有打开的界面就返回
    if self.panelStack:Count() == 1 then
        return
    else
        --关闭栈顶元素
        self.panelStack:Peek():CloseView()
        --退出已打开列表 出栈
        TableUtil.RemoveValue(self.panelList,self.panelStack:Peek().name)
        self.panelStack:Pop()

        --下方元素可见
        if self.panelStack:Count() ~=0 then
            self.panelStack:Peek():ShowView()
        end
    end
    
end